### ISSUE TITLE
unit cell esd determination should be folded into main xia2

### ISSUE THREAD

**Issue by Anthchirp** [xia2/xia2#28]
including:
* philosophical discussion of what a _correct unit cell_ is
* consistent handling of unit cell dimensions for .CIF and .INS file purposes
* writing information to xia2.cif

**Comment by graeme-winter** [comment]
Will discuss with @dagewa

**Comment by graeme-winter** [comment]
"Discuss"

> A philosophical question if I may to start the day
> 
> What is the “true” unit cell for a data set?
> 
> Right now this is a little bit of a head scratcher for xia2 running in small molecule mode because the logic is currently to use the unit cell from the scan static refinement as the “unit cell” but then I think we actually report the average of the scan varying unit cells weighted by #reflections in the MTZ output (and maybe in xia2 output)
> 
> It then gets a little more complicated when we consider small molecule land where people want *one* unit cell and *one* set of errors for 4 scans. We sorted this by Markus implementing a kind of monte-carlo pseudo powder refinement from the integrated reflections which seems to work OK however results in a slightly different result (by around 1e-14m) but is perhaps kinda sorta more “true” since it is not about modelling the spot positions.
> 
> So by now we have three conflicting definitions of the “right” unit cell is, any idea which one is “true”

**Comment by dagewa** [comment]
The static unit cell is the best least squares estimate of a single unit
cell that fits all the observed data included in refinement. "Best" in this
sense means it is the estimate that minimises the variance. This is related
to a strict definition of the "best" estimator (see BLUE - Best Linear
Unbiased Estimator), but we can't be that strict here, because the
definition requires that errors are uncorrelated. What we can say though,
is that by finding the minimum of the target function, the static cell
provides a single set of cell constants that minimises the deviation
between observations and predictions and produces concomitantly small error
estimates.

Ok, that sounds good, but what the semi-synthetic analysis in the
refinement paper showed is that although the error estimates thus produced
are a good estimate of the real error expected from purely random noise in
the centroid positions, the actual unit cell values can deviate from
expected values by many times the reported error. The case shown is a P 1
refinement of a truly P 4 cell where for example the alpha angle is
reported as 89.98317 +/- 0.00022. Clearly, systematic error in the centroid
positions can skew the accuracy of the highly precise result. [Aside: one
thing I ought to do now is go back and redo this analysis after correcting
some of that systematic error by recycling centroids after integration.
That process produced more believable scan-varying missets after all so one
might expect it to be a less biased data set].

There is also potentially a problem with the correctness of the centroid
variances going into refinement, which affect the weights. We have already
shown that these variance estimates are incorrect. This breaks the
assumption of the "statistical" weighting scheme in refinement. However,
when I last looked at this I think the results were still inaccurate on the
scale of the reported precision even when the estimated centroid variances
were replaced by real observed variances, i.e. the real scatter of
centroids over the 30 semi-synthetic datasets. So that is probably not the
biggest problem.

Despite the caveats, the static cell from refinement remains the best (most
powerful) predictive static model for the observed data. A scan-varying
average, like the Pointless/Aimless cell does not have the same predictive
power because if you plugged that cell into a static prediction model you
would end up with higher residuals. A scan-varying average cell has not
"seen" the centroid data in a refinement procedure so it has become
divorced from that target of reducing the mean square residual.

But the question is not what is the "best" cell, but what is the "true"
cell? Well, I don't know. The truth is unknown and the only tool we have to
get at the truth is a model fitting procedure. There are strictly-defined
model fitting procedures, like least squares minimisation, and other model
fitting procedures like post-processing minimisation results using an
averaging technique or inflation of errors to better fit prior
expectations. Also, is the true cell the cell that best represents the
recorded data set, or the cell that best represents the crystal? While you
probably want the first here, for comparison between cells (like in BLEND)
it is the second that might be more appropriate. Ideally that would require
a zero-dose extrapolation for the cell in order to come up with an estimate
of what it looked like before we started observing it.

If you decide to replace "truth" with "best" and define best as the static
model with greatest predictive power, and in turn define greatest
predictive power as producing a model with minimum variances, and then
additionally accept that the errors only reflect random noise and not any
systematic biases present in the data... Well, then the least squares
refined static unit cell and its error estimates are your "best" bet.
Producing one unit cell and one set of errors for 4 scans is feasible using
joint refinement. You will probably find you get a cell with errors that
"look too small" because if you took another crystal out and repeated the
experiment you would get quite different results. Yep, afraid so. The
errors you get from refinement are just those you would expect if you could
wind back time perfectly and repeat exactly the same experiment with
different quantum effects.

-- David

**Comment by rjgildea** [comment]
My main concern with this argument is that if you have done scan-varying refinement and allowed the experimental geometry to refine (as is usually the case) then the scan-static unit cell is no longer necessarily the "best" static model with the greatest predictive power for that particular set of experimental geometry parameters, as if you did a further scan-static refinement fixing the geometry you would undoubtedly get a different scan-static model to that which you had before scan-varying refinement.

**Comment by dagewa** [comment]
Yes, you can do exactly that. Refine a static cell using a fixed beam and detector purely for the purposes of determining the least squares average cell and its errors. All the caveats regarding "best" and "truth" still apply just as much.

**Comment by graeme-winter** [comment]
OK thanks folks did you reach a consensus on this? I am really of the opinion that getting something really "right" is close to impossible since everything we have one way or another is a model, just trying to get the most defensible model

**Comment by graeme-winter** [comment]
@dagewa can we yet to the fixed cell varying misset refinement we discussed back in the day? that would be *awesome* for this

**Comment by dagewa** [comment]
Do you mean

dials.refine refined.pickle refined_experiments.json scan_varying=true
crystal.fix=cell

?

This works

-- David

**Comment by dagewa** [comment]
Ah I guess you mean statically-refined cell with scan-varying angles?

-- David

**Comment by graeme-winter** [comment]
I mean scan varying U matrix but fixed B matrix - does the spell above do this for > 1 sweep?

**Comment by dagewa** [comment]
Certainly should do. There will be no reported errors on the B matrix though, as it is not parameterised

**Comment by graeme-winter** [comment]
Report from conversation on this subject:

The crntroid two-theta values contain *all* of the information needed to refine the unit cell (in @dagewa opinion) so therefore doing pseudo powder refinement is almost certainly a good way of deriving the best unit cell for the n scans together. Assert that this may be further improved by doing a proper least squares minimization & applying error propagation to derive "true" errors in place of the monte carlo approach currently implemented.

**Comment by dagewa** [comment]
It would be interesting to compare the unit cell obtained via the current
pseudo powder refinement method with a joint refinement of n scans of a
single crystal with everything else fixed, i.e. fixing also the crystal
orientations. The same information should be used to construct the dataset,
so if the powder rings are constructed from integrated centroids, then
centroid refinement should use integrated centroids too. I have a feeling
that they will be very similar, but I'd like to know by how much. It would
then also be interesting to compare the error estimates. This would be
trivially easy to check. If someone can provide me with a suitable dataset,
I will look at this.

**Comment by Anthchirp** [comment]
logic currently lives in xia2/Modules/UnitCellErrors.py

**Comment by dagewa** [comment]
Have you got a data set of the type you are most interested in (single crystal with multiple scans, centroids after integration) that I could use for a test? One that could live in dials_regression?

**Comment by graeme-winter** [comment]
Yes, I will upload something RSN unless @Anthchirp already doing?

**Comment by graeme-winter** [comment]
cd dials_regression && svn update:

```
Graemes-MacBook-Pro-4:dials_regression graeme$ svn commit -m "Raw material for xia2-28 https://github.com/xia2/xia2/issues/28"
Sending        doc/publications/dials_implementing_new_integration/implementing_evaluating_package.tex
Adding         xia2-28
Adding  (bin)  xia2-28/20_integrated.pickle
Adding         xia2-28/20_integrated_experiments.json
Adding  (bin)  xia2-28/25_integrated.pickle
Adding         xia2-28/25_integrated_experiments.json
Adding  (bin)  xia2-28/30_integrated.pickle
Adding         xia2-28/30_integrated_experiments.json
Adding  (bin)  xia2-28/35_integrated.pickle
Adding         xia2-28/35_integrated_experiments.json
Transmitting file data .........
Committed revision 1425.
```

**Comment by graeme-winter** [comment]
@Anthchirp  probably worth adding a little script which performs the same function as xia2.get_unit_cell_errors from these data without the need for all that messing around with xia2.jsons?

**Comment by graeme-winter** [comment]
Watching the ol' commits to DIALS looks like most of this is done already :)

**Comment by graeme-winter** [comment]
```
Graemes-MacBook-Pro-4:xia2-28 graeme$ dials.two_theta_refine [0-9]*
Sorry: Only one reflections list can be imported at present
```

sad face

```
Graemes-MacBook-Pro-4:xia2-28 graeme$ dials.two_theta_refine 20*
...
----------------------------
| Parameter | Estimated sd |
----------------------------
| 5.422     | 5.973e-05    |
| 8.1378    | 0.00010435   |
| 12.016    | 0.00014948   |
| 90        | 2.1785e-19   |
| 90        | 1.4558e-19   |
| 90        | 1.3526e-19   |
----------------------------
Saving refined experiments to refined_cell.json
```

happy face - damn that is fast

**Comment by graeme-winter** [comment]
Interesting - systematic two-theta offset?

```
Summary statistics for 1850 observations matched to predictions:
-----------------------------------------------------------------------------------------
|                           | Min       | Q1        | Med       | Q3        | Max       |
-----------------------------------------------------------------------------------------
| 2theta_c - 2theta_o (deg) | -0.6378   | -0.4916   | -0.4074   | -0.3171   | -0.0515   |
| 2theta weights            | 0.0003046 | 0.0003046 | 0.0003046 | 0.0003046 | 0.0003046 |
-----------------------------------------------------------------------------------------
```

#surprised

**Comment by graeme-winter** [comment]
506  dials.combine_experiments *pickle *json
  507  dials.two_theta_refine combined_*

:)

**Comment by graeme-winter** [comment]
```
----------------------------
| Parameter | Estimated sd |
----------------------------
| 5.428     | 9.5847e-05   |
| 8.1441    | 0.00014996   |
| 12.04     | 0.00021587   |
| 90        | 3.1383e-19   |
| 90        | 2.1314e-19   |
| 90        | 2.1243e-19   |
----------------------------
Saving refined experiments to refined_cell.json

```

**Comment by graeme-winter** [comment]
For i04-bag-training data no bias observed in two-theta input; for dna data measured with wavelength 0.551040 get

```
-----------------------------------------------------------------------------------------
|                           | Min       | Q1        | Med       | Q3        | Max       |
-----------------------------------------------------------------------------------------
| 2theta_c - 2theta_o (deg) | -0.04701  | -0.01462  | -0.006991 | 0.0005607 | 0.03353   |
| 2theta weights            | 0.0003046 | 0.0003046 | 0.0003046 | 0.0003046 | 0.0003046 |
-----------------------------------------------------------------------------------------
```

i.e. very small systematic offset

Refinement steps small

```
Refinement steps:
-------------------------------
| Step | Nref   | RMSD_2theta |
|      |        | (deg)       |
-------------------------------
| 0    | 103968 | 0.013265    |
| 1    | 103968 | 0.010858    |
| 2    | 103968 | 0.010857    |
-------------------------------
```

**Comment by graeme-winter** [comment]
Setting the spacegroup to P1 seems to do "the right thing" though the better way @rjgildea says is to compute the correct reindex to P1, apply this then apply the inverse to the refined cell.

**Comment by graeme-winter** [comment]
@dagewa to integrate combine_experiments into the command interpretation => can add multiple json / pickle inputs

**Comment by graeme-winter** [comment]
This should be used always to give the final unit cell for xia2 i.e. with small molecule and big molecule data if we used dials.

**Comment by graeme-winter** [comment]
Hmmm.... while we are here also seems to make a lot of difference to applying the refinement after integration using the good centroids i.e. filter for strong && integrated gives *fantastic* improvement in RMSD:

```
------------------------------------------------
| Step | Nref | RMSD_X   | RMSD_Y   | RMSD_Phi |
|      |      | (mm)     | (mm)     | (deg)    |
------------------------------------------------
| 0    | 1234 | 0.85004  | 1.2543   | 0.18753  |
| 1    | 1234 | 0.12486  | 0.13739  | 0.17553  |
| 2    | 1234 | 0.11399  | 0.12801  | 0.16604  |
| 3    | 1234 | 0.099415 | 0.11186  | 0.14233  |
| 4    | 1234 | 0.075066 | 0.079777 | 0.09627  |
| 5    | 1234 | 0.054283 | 0.046631 | 0.044695 |
| 6    | 1234 | 0.050306 | 0.037329 | 0.025557 |
| 7    | 1234 | 0.050357 | 0.036943 | 0.024515 |
| 8    | 1234 | 0.050373 | 0.036937 | 0.024499 |
| 9    | 1234 | 0.050373 | 0.036937 | 0.024498 |
------------------------------------------------
RMSD no longer decreasing

RMSDs by experiment:
---------------------------------------------
| Exp | Nref | RMSD_X  | RMSD_Y  | RMSD_Z   |
| id  |      | (px)    | (px)    | (images) |
---------------------------------------------
| 0   | 1234 | 0.29287 | 0.21475 | 0.24498  |
---------------------------------------------
```

Should this also be a feature of xia2 i.e. global refinement after processing & reintegration c/f old deadly

**Comment by Anthchirp** [comment]
Looking through xia2.get_unit_cell_error code. [There I reindex every sweep](https://github.com/xia2/xia2/blob/master/command_line/get_unit_cell_errors.py#L95), so they all end up with same setting. Don't see similar logic in dials.two_theta_refine. Would that be redundant?
I remember this was important for some CX dataset, but don't remember a particular one.

**Comment by graeme-winter** [comment]
Not needed for two theta refine, I think... should mention to @dagewa though

**Comment by Anthchirp** [comment]
@dagewa two more questions:
For the .CIF I need the minimum/maximum used HKL indices in refinement. From what I understand these are currently [all miller indices](https://github.com/dials/dials/commit/60f88ba5c1ebbc8c5047d4ffa5324c3eda2e3aef#diff-0d70fc287715a36436f8a705930f4c81R309), ie. there is no filtering?
On a similar note I also need the minimum and maximum (two)theta angle of the reflections used. Is there an easy/canonical way to access eg reflections['2theta_obs.rad']?

**Comment by dagewa** [comment]
@Anthchirp Re: reindexing - it shouldn't matter, because only the 2theta angle is predicted and compared with observed.

For the other two questions, do you mean to access this information from within two_theta_refine, or should it be writing the reflections used in refinement to disk?

**Comment by dagewa** [comment]
From within two_theta_refine, you can use 

`rt=refiner.get_matches()`

to obtain the reflection table actually used in refinement. Then `flex.min(rt['2theta_obs.rad'])` etc. seems reasonable.

**Comment by dagewa** [comment]
BTW there _may_ be some filtering of reflections. Certainly any junk like 0,0,0 gets removed, though shouldn't be an issue if the input came from integration. There is also the standard Refiner's too-close-to-spindle criterion applied. In the case of normal refinement this is to avoid derivatives blowing up to infinity. For two_theta_refine that isn't relevant, but it does ensure that the same set of reflections are accepted by default settings by this and dials.refine. It is likely that these reflections don't exist anyway in integrated.pickle though, but using `refiner.get_matches()` is sensible to avoid any uncertainty.

**Comment by Anthchirp** [comment]
Don't need the actual numbers on disk, but do need some summary statistics for the CIF. I think/hope I got the right numbers, but will validate against xia2.get_unit_cell_errors to be sure.

**Comment by graeme-winter** [comment]
In discussion this looks in a fit state to pull in.
